dex/wait/queue.go

```diff
@@ -102,3 +111,135 @@ out:
+func NewTaperingTickerQueue(fastestInterval, slowestInterval time.Duration) *TaperingTickerQueue {
+	return &TaperingTickerQueue{
+		waiters:         make([]*taperingWaiter, 0, 100),
+		fastestInterval: fastestInterval,
+		slowestInterval: slowestInterval,
+		recalcTimer:     make(chan struct{}, 1),
+	}
+}
+
+// Wait attempts to run the (*Waiter).TryFunc until either 1) the function
+// returns the value DontTryAgain, or 2) the function's Expiration time has
+// passed. In the case of 2, the (*Waiter).ExpireFunc will be run.
+func (q *TaperingTickerQueue) Wait(waiter *Waiter) {
+	if time.Now().After(waiter.Expiration) {
+		log.Error("wait.TickerQueue: Waiter given expiration before present")
+		return
+	}
+	// We don't want the caller to hang here, so we won't call TryFunc. Instead
+	// set the nextTick as now and the run loop will call it in a goroutine
+	// immediately.
+	q.insert(&taperingWaiter{Waiter: waiter}, time.Now())
+}
+
+// insert inserts the waiter, setting its nextTick under waiterMtx lock, and
+// sorts the waiters by time until nextTick.
+func (q *TaperingTickerQueue) insert(w *taperingWaiter, nextTick time.Time) {
+	q.waiterMtx.Lock()
+	w.nextTick = nextTick
+	q.waiters = append(q.waiters, w)
+	sort.Slice(q.waiters, func(i, j int) bool { return q.waiters[i].nextTick.Before(q.waiters[j].nextTick) })
+	q.waiterMtx.Unlock()
+	q.recalcTimer <- struct{}{}
+}
+
+// Run runs the primary wait loop until the context is canceled.
+func (q *TaperingTickerQueue) Run(ctx context.Context) {
+
+	taper := float64(q.slowestInterval - q.fastestInterval)
+
+	runWaiter := func(w *taperingWaiter, n int) {
+		if w.TryFunc() == DontTryAgain {
+			return
+		}
+		// If this waiter has expired, issue the timeout error to the client
+		// and don't re-insert.
+		if w.Expiration.Before(time.Now()) {
+			w.ExpireFunc()
+			return
+		}
+
+		var nextTick time.Time
+		switch {
+		case n <= fullSpeedTicks:
+			nextTick = time.Now().Add(q.fastestInterval)
+		case n < fullyTapered: // ramp up the interval
+			prog := float64(n-fullSpeedTicks) / (fullyTapered - fullSpeedTicks)
+			interval := q.fastestInterval + time.Duration(math.Round(prog*taper))
+			nextTick = time.Now().Add(interval)
+		default:
+			nextTick = time.Now().Add(q.slowestInterval)
+		}
+		if nextTick.After(w.Expiration) {
+			nextTick = w.Expiration
+		}
+		q.insert(w, nextTick)
+	}
+
+	for {
+		q.waiterMtx.Lock()
+		var nextTick <-chan time.Time
+		if len(q.waiters) > 0 {
+			nextTick = time.After(time.Until(q.waiters[0].nextTick))
+		}
+		q.waiterMtx.Unlock()
+
+		select {
+		case <-nextTick:
+			q.waiterMtx.Lock()
+			// No need to check length. This loop is the only place waiters can
+			// be removed from the slice and it's sychronous.
+			w := q.waiters[0]
+			tick := w.tick
+			w.tick++
+			// Remove the waiter from the slice. runWaiter will re-insert if
+			// it sees TryAgain.
+			q.waiters = q.waiters[1:]
+			q.waiterMtx.Unlock()
+			go runWaiter(w, tick)
+		case <-q.recalcTimer:
```

Comment:
This event loop design actually lends itself very nicely to a lock-free design, which you were very close to having.  Consider: https://github.com/buck54321/dcrdex/compare/better-latency-handling...chappjc:better-latency-handling-chappjc?expand=1

Reply:
I wanted to even make the `waiters` slice local to `Run`, since that's the essence of the change, but didn't want to break that last check in your test.